### PR TITLE
Upgrade from Storybook v7 -> Storybook v8

### PR DESCRIPTION
**Pull request by jesstelford**
About 90% of the upgrade was handled by `npx storybook@latest upgrade` 🎉  _(Feels like just [yesterday we did the upgrade from v6 -> v7](https://github.com/Shopify/polaris/pull/9019) 😅)_

## Node v18

Storybook v8 requires node v18. But we're still stuck on node v16.

🚨 This PR is blocked by;
- ~https://github.com/Shopify/polaris/pull/11844~
- ~https://github.com/Shopify/polaris/pull/11924~
- ~https://github.com/Shopify/polaris/pull/11930~

## Performance

Calculated using [hyperfine](https://github.com/sharkdp/hyperfine):

```sh
hyperfine --parameter-list branch main,storybook-8 --setup='git checkout {branch} && yarn && yarn build --filter=@shopify/polaris' --prepare='rm -rf polaris-react/build-internal/storybook/' --warmup 1 'yarn workspace @shopify/polaris run storybook:build --quiet'
```

**`main`**

```
Benchmark 1: yarn workspace @shopify/polaris run storybook:build --quiet
  Time (mean ± σ):     24.362 s ±  0.387 s    [User: 39.938 s, System: 4.076 s]
  Range (min … max):   23.896 s … 25.034 s    10 runs

```

**`storybook-8`**

```
Benchmark 2: yarn workspace @shopify/polaris run storybook:build --quiet
  Time (mean ± σ):     29.486 s ±  1.321 s    [User: 40.765 s, System: 3.868 s]
  Range (min … max):   28.419 s … 32.763 s    10 runs
```

**Results**

```
Summary
  yarn workspace @shopify/polaris run storybook:build --quiet (branch = main) ran
    1.21 ± 0.06 times faster than yarn workspace @shopify/polaris run storybook:build --quiet (branch = storybook-8)
```

Combined with the slow-down seen in [the upgrade to v7](https://github.com/Shopify/polaris/pull/9019), I'm starting to think Chromatic's marketing about better performance isn't true for us 😅

### CODE REVIEW

**Comment by BPScott** [comment]
I'm kicking this SB8 update over in checkout-web and found a storybook-a11y-test patch that works for SB8 (I've not tested if it falls back well in SB7/SB6) over in: https://github.com/Shopify/checkout-web/compare/sb8?expand=1

**Review by alex-page — APPROVED** [review]

File: polaris-react/scripts/accessibility-check.js [hunk 1]
```diff
@@ -1,40 +0,0 @@
-/* eslint-disable no-console */
```
Comment (jesstelford) [inline polaris-react/scripts/accessibility-check.js:1]: Replaced with `polaris-react/.storybook/test-runner.ts`

File: polaris-react/package.json [hunk 2]
```diff
@@ -52,7 +52,8 @@
     "clean": "rm -rf .turbo node_modules build build-internal",
     "start": "serve ./build-internal/storybook/static -l ${POLARIS_STORYBOOK_PORT:=6006}",
     "storybook": "storybook dev -p ${POLARIS_STORYBOOK_PORT:=6006} --quiet",
-    "storybook:build": "storybook build -o build-internal/storybook/static"
+    "storybook:build": "storybook build -o build-internal/storybook/static",
+    "storybook:test": "concurrently -k -s first -n 'SB,TEST' -c 'magenta,blue' 'http-server build-internal/storybook/static --port 6006 --silent' 'wait-on tcp:6006 && test-storybook --maxWorkers=2 polaris-react/src/components/Modal/Modal.stories.tsx'"
```
Comment (jesstelford) [inline polaris-react/package.json:56]: This command is from the Storybook docs

File: polaris-react/playground/DetailsPage.tsx [hunk 3]
```diff
@@ -52,6 +52,7 @@ import type {DropZoneProps, PageProps} from '../src';
 import styles from './DetailsPage.module.css';
 
 export const DetailsPage = {
+  tags: ['skip-tests'],
```
Comment (jesstelford) [inline polaris-react/playground/DetailsPage.tsx:55]: This used to be in `polaris-react/scripts/accessibility-check.js`

File: polaris-react/src/components/Banner/Banner.stories.tsx [hunk 4]
```diff
@@ -142,6 +142,7 @@ export const InAModal = {
       <div style={{height: '500px'}}>
         <Button onClick={handleChange}>Open</Button>
         <Modal
+          instant
```
Comment (jesstelford) [inline polaris-react/src/components/Banner/Banner.stories.tsx:145]: Prevents the accessibility checker from running while the CSS fade-in transition for Modal is still happening and giving false positives for color contrast issues.

File: polaris-react/src/components/Modal/Modal.stories.tsx [hunk 5]
```diff
@@ -15,12 +16,16 @@ import {
...
 } from '@shopify/polaris';
+import {transitionsAllSettled} from 'transitions-all-settled';
 
 export default {
   component: Modal,
 } as Meta<typeof Modal>;
 
 export const Default = {
+  play: async ({canvasElement}) => {
+    await transitionsAllSettled(canvasElement);
```
Comment (jesstelford) [inline polaris-react/src/components/Modal/Modal.stories.tsx:27]: Prevents the accessibility checker from running while the CSS fade-in transition for Modal is still happening and giving false positives for color contrast issues.

File: polaris-react/.storybook/test-runner.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
+/* eslint-env node */
```
Comment (jesstelford) [inline polaris-react/.storybook/test-runner.ts:1]: 🙏 🙇 Thankyou @BPScott for [pointing me in the right direction](.changeset/rich-planets-allow.md) with this one!
🤜 🤛

**Pull request merged**